src/Stratis.Bitcoin.Features.Dns/UnreliablePeerBehavior.cs

```diff
@@ -0,0 +1,136 @@
+        private readonly IPeerBanning peerBanning;
+
+        /// <summary>The node settings.</summary>
+        private readonly NodeSettings nodeSettings;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public UnreliablePeerBehavior(Network network, IChainState chainState, ILoggerFactory loggerFactory, IPeerBanning peerBanning, NodeSettings nodeSettings)
+        {
+            Guard.NotNull(network, nameof(network));
+            Guard.NotNull(chainState, nameof(chainState));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(peerBanning, nameof(nodeSettings));
+            Guard.NotNull(nodeSettings, nameof(nodeSettings));
+
+            this.network = network;
+            this.chainState = chainState;
+            this.loggerFactory = loggerFactory;
+            this.peerBanning = peerBanning;
+            this.nodeSettings = nodeSettings;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public override object Clone()
+        {
+            return new UnreliablePeerBehavior(this.network, this.chainState, this.loggerFactory, this.peerBanning, this.nodeSettings);
+        }
+
+        /// <inheritdoc />
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync);
+        }
+
+        /// <inheritdoc />
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceivedAsync);
+        }
+
+        /// <summary>
+        /// Event handler that is called when the node receives a network message from the attached peer.
+        /// </summary>
+        /// <param name="peer">Peer that sent us the message.</param>
+        /// <param name="message">Received message.</param>
+        /// <remarks>
+        /// This handler only cares about "version" messages, which are only sent once per node.
+        /// </remarks>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            if (message.Message.Payload is VersionPayload version)
+            {
+                // If current node is on POS, and ProvenHeaders is activated, check if current connected peer can serve Proven Headers.
+                // If it can't, disconnect from him and ban for few minutes
+                if (IsProvenHeaderActivated())
+                {
+                    if (!CanServeProvenHeader(version))
+                    {
+                        TimeSpan banDuration = TimeSpan.FromMinutes(1);
+                        this.logger.LogDebug("Peer '{0}' has been disconnected for {1} because can't serve proven headers. Peer Version: {2}", peer.RemoteSocketEndpoint, banDuration, version.Version);
```

Comment:
We shouldn't ban the peer, just add it to some kind of collection against which whitelist manager will check peers before adding them to masterfile.

Reply:
@dangershony told me to disconnect peers that were not able to serve it, I banned them for few minutes because I didn't knew if just disconnecting were causing them to try to reconnect again.

I can remove the ban and just disconnect (this mean I can remove the `IPeerBanning` reference too from that class, but I'd disconnect them as per @dangershony request, or if you agree on put them on some list I'd have to refactor to exclude those peers in the 
https://github.com/statisproject/StratisBitcoinFullNode/blob/51ab318a001158b268ee398a5bfeb902da3d7ac2/src/Stratis.Bitcoin.Features.Dns/WhitelistManager.cs#L100

Not a problem, I've just to know what to do.

Reply:
Why would we disconnect those? It will mean that old nodes won't be able to use DNS